parser-typechecker/src/Unison/CommandLine/DisplayValues.hs

```diff
@@ -0,0 +1,102 @@
+import Unison.Term (AnnotatedTerm)
+import Unison.Type (Type)
+import Unison.Var (Var)
+import qualified Unison.DataDeclaration as DD
+import qualified Unison.DeclPrinter as DP
+import qualified Unison.NamePrinter as NP
+import qualified Unison.PrettyPrintEnv as PPE
+import qualified Unison.Referent as Referent
+import qualified Unison.Reference as Reference
+import qualified Unison.Term as Term
+import qualified Unison.TermPrinter as TP
+import qualified Unison.TypePrinter as TypePrinter
+import qualified Unison.Util.Pretty as P
+import qualified Unison.Util.SyntaxText as S
+
+type Pretty = P.Pretty P.ColorText
+
+displayTerm :: (Var v, Monad m)
+           => PPE.PrettyPrintEnv 
+           -> (Reference -> m (Maybe (AnnotatedTerm v a)))
+           -> (Referent -> m (Maybe (Type v a)))
+           -> (Reference -> m (Maybe (AnnotatedTerm v a)))
+           -> (Reference -> m (Maybe (DD.Decl v a)))
+           -> AnnotatedTerm v a 
+           -> m Pretty
+displayTerm ppe terms typeOf eval types tm = case tm of
+  -- todo: can dispatch on other things with special rendering
+  Term.Ref' r -> eval r >>= \case
+    Nothing -> pure $ termName ppe (Referent.Ref r) 
+    Just tm -> displayDoc ppe terms typeOf eval types tm
+  _ -> displayDoc ppe terms typeOf eval types tm
+
+displayDoc :: (Var v, Monad m)
+           => PPE.PrettyPrintEnv 
+           -> (Reference -> m (Maybe (AnnotatedTerm v a)))
+           -> (Referent  -> m (Maybe (Type v a)))
+           -> (Reference -> m (Maybe (AnnotatedTerm v a)))
+           -> (Reference -> m (Maybe (DD.Decl v a)))
+           -> AnnotatedTerm v a 
+           -> m Pretty
+displayDoc ppe terms typeOf evaluated types t = go t
+  where
+  go (DD.DocJoin docs) = foldMap id <$> traverse go docs
+  go (DD.DocBlob txt) = pure $ P.paragraphyText txt
+  go (DD.DocLink (DD.LinkTerm (Term.TermLink' r))) = pure $ termName ppe r 
+  go (DD.DocLink (DD.LinkType (Term.TypeLink' r))) = pure $ typeName ppe r 
+  go (DD.DocSource (DD.LinkTerm (Term.TermLink' r))) = prettyTerm terms r 
+  go (DD.DocSource (DD.LinkType (Term.TypeLink' r))) = prettyType r
+  go (DD.DocSignature (Term.TermLink' r)) = prettySignature r
+  go (DD.DocEvaluate (Term.TermLink' r)) = prettyTerm evaluated r
+  go tm = pure $ TP.pretty ppe tm
+  prettySignature r = typeOf r >>= \case
+    Nothing -> pure $ termName ppe r 
+    Just typ -> pure $ TypePrinter.prettySignatures ppe [(PPE.termName ppe r, typ)]
+  prettyTerm terms r = case r of
+    Referent.Ref (Reference.Builtin _) -> prettySignature r 
+    Referent.Ref ref -> terms ref >>= \case
+      Nothing -> pure $ "😶  Missing term source for: " <> termName ppe r
+      Just tm -> pure . P.syntaxToColor $ TP.prettyBinding ppe (PPE.termName ppe r) tm
+    Referent.Con r _ _ -> prettyType r 
+  prettyType r = types r >>= \case
+    Nothing -> pure $ "😶  Missing type source for: " <> typeName ppe r
+    Just ty -> pure . P.syntaxToColor $ DP.prettyDecl ppe r (PPE.typeName ppe r) ty
```

Comment:
This is the function responsible for rendering a `Doc`. Uses the pattern synonyms, so this is actually quite readable.